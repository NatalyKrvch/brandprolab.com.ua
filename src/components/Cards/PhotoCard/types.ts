export interface PhotoCardProps {
  photoUrl: string;
  cardHeight: number;
  imageWidth: number;
  imageHeight: number;
  backgroundUrl: string;
  text?: string;
}
