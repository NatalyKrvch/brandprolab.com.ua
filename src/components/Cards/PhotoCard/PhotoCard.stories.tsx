import type { Meta, StoryObj } from '@storybook/react';

import PhotoCard from './PhotoCard';

const meta: Meta<typeof PhotoCard> = {
  title: 'Components/PhotoCard',
  component: PhotoCard,
  parameters: {
    controls: { expanded: true },
  },
  argTypes: {
    photoUrl: {
      control: 'text',
      description: 'URL of the photo to display.',
      table: {
        type: { summary: 'string' },
      },
    },
    backgroundUrl: {
      control: 'text',
      description: 'URL of the background image.',
      table: {
        type: { summary: 'string' },
      },
    },
    imageWidth: {
      control: 'number',
      description: 'Width of the photo.',
      table: {
        type: { summary: 'number' },
      },
    },
    imageHeight: {
      control: 'number',
      description: 'Height of the photo.',
      table: {
        type: { summary: 'number' },
      },
    },
    cardHeight: {
      control: 'number',
      description: 'Height of the card.',
      table: {
        type: { summary: 'number' },
      },
    },
    text: {
      control: 'text',
      description: 'Optional text displayed on the card.',
      table: {
        type: { summary: 'string' },
      },
    },
  },
};

export default meta;
type Story = StoryObj<typeof PhotoCard>;

export const WithText: Story = {
  args: {
    text: 'Чому мені можна довіряти?',
    photoUrl: '/images/temp/photo-expert.png',
    backgroundUrl: '/images/backgrounds/expert-bg.svg',
    imageWidth: 362,
    imageHeight: 600,
    cardHeight: 744,
  },
};

export const WithoutText: Story = {
  args: {
    photoUrl: '/images/temp/photo-hero.png',
    backgroundUrl: '/images/backgrounds/hero-photocard-bg.svg',
    imageWidth: 512,
    imageHeight: 764,
    cardHeight: 824,
  },
};
