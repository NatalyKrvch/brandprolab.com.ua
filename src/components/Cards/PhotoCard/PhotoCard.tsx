import Image from 'next/image';

import {
  PHOTO_CARD_IMAGE_TEST_ID,
  PHOTO_CARD_TEXT_TEST_ID,
} from '@/lib/testIDs';

import { SimpleCard } from '../SimpleCard';
import type { PhotoCardProps } from './types';

const PhotoCard = ({
  photoUrl,
  backgroundUrl,
  cardHeight,
  imageWidth,
  imageHeight,
  text,
  photoAlt = 'Photo',
  hasPriority = false,
}: PhotoCardProps) => {
  return (
    <SimpleCard
      backgroundImageUrl={backgroundUrl}
      className="relative"
      style={{ height: `${cardHeight}px` }}
    >
      {text && (
        <div
          className="absolute left-4 top-4 ml-36 mt-44 w-80 text-44 font-bold leading-44 tracking-tight text-white"
          data-testid={PHOTO_CARD_TEXT_TEST_ID}
        >
          {text}
        </div>
      )}

      <div
        className="absolute bottom-0 left-1/2 -translate-x-1/2"
        style={{ width: `${imageWidth}px`, height: `${imageHeight}px` }}
      >
        <Image
          src={photoUrl}
          fill
          className="object-cover"
          alt={photoAlt}
          priority={hasPriority}
          data-testid={PHOTO_CARD_IMAGE_TEST_ID}
        />
      </div>
    </SimpleCard>
  );
};

export default PhotoCard;
