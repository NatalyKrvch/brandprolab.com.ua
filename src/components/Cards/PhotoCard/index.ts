export { default as PhotoCard } from './PhotoCard';
export * from './types';
