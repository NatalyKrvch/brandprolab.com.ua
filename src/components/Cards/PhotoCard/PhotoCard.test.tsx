import '@testing-library/jest-dom';

import { render, screen } from '@testing-library/react';

import {
  PHOTO_CARD_IMAGE_TEST_ID,
  PHOTO_CARD_TEXT_TEST_ID,
  SIMPLE_CARD_TEST_ID,
} from '@/lib/testIDs';

import PhotoCard from './PhotoCard';

jest.mock('next/image', () => {
  const MockImage = ({ src, alt }: { src: string; alt: string }) => (
    <img src={src} alt={alt} data-testid={PHOTO_CARD_IMAGE_TEST_ID} />
  );
  MockImage.displayName = 'MockImage';
  return MockImage;
});

describe('PhotoCard component', () => {
  const defaultProps = {
    text: 'Sample Text',
    photoUrl: '/test-photo.jpg',
    backgroundUrl: '/test-background.jpg',
  };

  it('renders the PhotoCard component', () => {
    const { asFragment } = render(<PhotoCard {...defaultProps} />);
    const cardElement = screen.getByTestId(SIMPLE_CARD_TEST_ID);

    expect(cardElement).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders text when provided', () => {
    const { asFragment } = render(<PhotoCard {...defaultProps} />);
    const textElement = screen.getByTestId(PHOTO_CARD_TEXT_TEST_ID);

    expect(textElement).toBeInTheDocument();
    expect(textElement).toHaveTextContent('Sample Text');
    expect(asFragment()).toMatchSnapshot();
  });

  it('does not render text when not provided', () => {
    const { asFragment } = render(
      <PhotoCard
        photoUrl={defaultProps.photoUrl}
        backgroundUrl={defaultProps.backgroundUrl}
      />,
    );
    expect(
      screen.queryByTestId(PHOTO_CARD_TEXT_TEST_ID),
    ).not.toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders the photo correctly', () => {
    const { asFragment } = render(<PhotoCard {...defaultProps} />);
    const imageElement = screen.getByTestId(PHOTO_CARD_IMAGE_TEST_ID);

    expect(imageElement).toBeInTheDocument();
    expect(imageElement).toHaveAttribute('src', '/test-photo.jpg');
    expect(imageElement).toHaveAttribute('alt', 'Photo');
    expect(asFragment()).toMatchSnapshot();
  });
});
