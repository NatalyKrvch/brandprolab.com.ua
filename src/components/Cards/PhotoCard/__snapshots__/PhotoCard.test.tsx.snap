// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhotoCard component applies correct styles for dimensions 1`] = `
<DocumentFragment>
  <div
    class="w-full bg-cover bg-center bg-white rounded-24 tablet:rounded-32 desktop:rounded-40 relative"
    data-testid="simple-card-test-id"
    style="height: 500px; background-image: url(/test-background.jpg); background-size: cover; background-position: center; background-repeat: no-repeat;"
  >
    <div
      class="absolute left-4 top-4 ml-36 mt-44 w-80 text-44 font-bold leading-44 tracking-tight text-white"
      data-testid="photo-card-text-test-id"
    >
      Sample Text
    </div>
    <div
      class="absolute bottom-0 left-1/2 -translate-x-1/2"
      style="width: 200px; height: 200px;"
    >
      <img
        alt="Photo"
        class="object-cover"
        data-testid="photo-card-image-test-id"
        src="/test-photo.jpg"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`PhotoCard component does not render text when not provided 1`] = `
<DocumentFragment>
  <div
    class="w-full bg-cover bg-center bg-white rounded-24 tablet:rounded-32 desktop:rounded-40 relative"
    data-testid="simple-card-test-id"
    style="height: 500px; background-image: url(/test-background.jpg); background-size: cover; background-position: center; background-repeat: no-repeat;"
  >
    <div
      class="absolute bottom-0 left-1/2 -translate-x-1/2"
      style="width: 200px; height: 200px;"
    >
      <img
        alt="Photo"
        class="object-cover"
        data-testid="photo-card-image-test-id"
        src="/test-photo.jpg"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`PhotoCard component renders text when provided 1`] = `
<DocumentFragment>
  <div
    class="w-full bg-cover bg-center bg-white rounded-24 tablet:rounded-32 desktop:rounded-40 relative"
    data-testid="simple-card-test-id"
    style="height: 500px; background-image: url(/test-background.jpg); background-size: cover; background-position: center; background-repeat: no-repeat;"
  >
    <div
      class="absolute left-4 top-4 ml-36 mt-44 w-80 text-44 font-bold leading-44 tracking-tight text-white"
      data-testid="photo-card-text-test-id"
    >
      Sample Text
    </div>
    <div
      class="absolute bottom-0 left-1/2 -translate-x-1/2"
      style="width: 200px; height: 200px;"
    >
      <img
        alt="Photo"
        class="object-cover"
        data-testid="photo-card-image-test-id"
        src="/test-photo.jpg"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`PhotoCard component renders the PhotoCard component 1`] = `
<DocumentFragment>
  <div
    class="w-full bg-cover bg-center bg-white rounded-24 tablet:rounded-32 desktop:rounded-40 relative"
    data-testid="simple-card-test-id"
    style="height: 500px; background-image: url(/test-background.jpg); background-size: cover; background-position: center; background-repeat: no-repeat;"
  >
    <div
      class="absolute left-4 top-4 ml-36 mt-44 w-80 text-44 font-bold leading-44 tracking-tight text-white"
      data-testid="photo-card-text-test-id"
    >
      Sample Text
    </div>
    <div
      class="absolute bottom-0 left-1/2 -translate-x-1/2"
      style="width: 200px; height: 200px;"
    >
      <img
        alt="Photo"
        class="object-cover"
        data-testid="photo-card-image-test-id"
        src="/test-photo.jpg"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`PhotoCard component renders the photo with correct attributes 1`] = `
<DocumentFragment>
  <div
    class="w-full bg-cover bg-center bg-white rounded-24 tablet:rounded-32 desktop:rounded-40 relative"
    data-testid="simple-card-test-id"
    style="height: 500px; background-image: url(/test-background.jpg); background-size: cover; background-position: center; background-repeat: no-repeat;"
  >
    <div
      class="absolute left-4 top-4 ml-36 mt-44 w-80 text-44 font-bold leading-44 tracking-tight text-white"
      data-testid="photo-card-text-test-id"
    >
      Sample Text
    </div>
    <div
      class="absolute bottom-0 left-1/2 -translate-x-1/2"
      style="width: 200px; height: 200px;"
    >
      <img
        alt="Photo"
        class="object-cover"
        data-testid="photo-card-image-test-id"
        src="/test-photo.jpg"
      />
    </div>
  </div>
</DocumentFragment>
`;
